src/basic/process-util.c

```diff
@@ -2054,14 +2060,57 @@ int posix_spawn_wrapper(const char *path, char *const *argv, char *const *envp,
         r = posix_spawnattr_init(&attr);
         if (r != 0)
                 return -r; /* These functions return a positive errno on failure */
-        /* Set all signals to SIG_DFL */
-        r = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGMASK|POSIX_SPAWN_SETSIGDEF);
+
+#if HAVE_PIDFD_SPAWN
+        if (cgroup) {
+                r = cg_get_path_and_check(
+                                SYSTEMD_CGROUP_CONTROLLER,
+                                cgroup,
+                                /* suffix= */ NULL,
+                                &resolved_cgroup);
+                if (r < 0) {
+                        r = -r;
+                        goto fail;
+                }
+
+                cgroup_fd = open(resolved_cgroup, O_PATH|O_DIRECTORY|O_CLOEXEC);
+                if (cgroup_fd < 0) {
+                        r = errno;
+                        goto fail;
+                }
+
+                r = posix_spawnattr_setcgroup_np(&attr, cgroup_fd);
+                if (r != 0)
+                        goto fail;
+
+                flags |= POSIX_SPAWN_SETCGROUP;
+        }
+#endif
+
+        r = posix_spawnattr_setflags(&attr, flags);
         if (r != 0)
                 goto fail;
         r = posix_spawnattr_setsigmask(&attr, &mask);
         if (r != 0)
                 goto fail;
 
+#if HAVE_PIDFD_SPAWN
+        r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+        if (r == 0) {
+                posix_spawnattr_destroy(&attr);
+                return pidref_set_pidfd_consume(ret_pidref, TAKE_FD(pidfd));
+        }
+        if (!(ERRNO_IS_NOT_SUPPORTED(errno) || ERRNO_IS_PRIVILEGE(errno)))
```

Comment:
how does pidfd_spawn() return its errors? 

if it's modelled after posix_spawn() then i'd expect the possitive errno is in `r`, not `errno`?

if it's more regularly modelled, then i'd figure the success check above should probably be `if (r >= 0) rather than `if (r == 0)`, simply to use the same pattern we use everywhere else.

Reply:
It happens to set errno when it gets ENOSYS, ie when I was testing the fallback path using seccomp. But it's the same implementation as posix_spawn, so for other cases it just returns the error as the return value, so I've switched it.